Sources/Build/BuildPlan.swift

```diff
@@ -1273,64 +1270,60 @@ public class BuildPlan {
     /// Diagnostics Engine for emitting diagnostics.
     let diagnostics: DiagnosticsEngine
 
-    private static func planLinuxMain(
+    private static func makeTestManifestTargets(
         _ buildParameters: BuildParameters,
         _ graph: PackageGraph
-    ) throws -> [(ResolvedProduct, SwiftTargetBuildDescription)] {
+    ) throws -> [(product: ResolvedProduct, targetBuildDescription: SwiftTargetBuildDescription)] {
         guard !buildParameters.triple.isDarwin() else {
             return []
         }
 
         var result: [(ResolvedProduct, SwiftTargetBuildDescription)] = []
-
         for testProduct in graph.allProducts where testProduct.type == .test {
-            // Create the target description from the linux main if test discovery is off.
-            if !buildParameters.enableTestDiscovery {
-                guard let linuxMainTarget = testProduct.linuxMainTarget else {
-                    throw Error.missingLinuxMain
-                }
-
+            // if test manifest exists, prefer that over test detection,
+            // this is designed as an escape hatch when test discovery is not appropriate
+            // and for backwards compatibility for projects that have existing test manifests (LinuxMain.swift)
+            if let testManifestTarget = testProduct.testManifestTarget {
                 let desc = try SwiftTargetBuildDescription(
-                    target: linuxMainTarget,
+                    target: testManifestTarget,
                     buildParameters: buildParameters,
                     isTestTarget: true
                 )
 
                 result.append((testProduct, desc))
-                continue
-            }
```

Comment:
the diff here is a bit confusing, it builds down to if/else instead of a subtle "continue"